File: sdk/typescript/src/client.test.ts
import { GradienceClient, GradienceError } from "./client";

describe("GradienceClient", () => {
  let client: GradienceClient;
  const fetchMock = jest.fn();

  beforeEach(() => {
    client = new GradienceClient("http://localhost:8080", { apiToken: "test-token" });
    global.fetch = fetchMock;
    fetchMock.mockClear();
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  function mockResponse(status: number, body: unknown) {
    fetchMock.mockResolvedValueOnce({
      ok: status < 400,
      status,
      json: async () => body,
      text: async () => JSON.stringify(body),
    } as Response);
  }

  it("creates a wallet", async () => {
    mockResponse(201, { id: "w1", name: "demo", status: "active", created_at: "2026-01-01T00:00:00Z" });
    const wallet = await client.createWallet("demo");
    expect(wallet.id).toBe("w1");
    expect(wallet.name).toBe("demo");
    expect(fetchMock).toHaveBeenCalledWith(
      "http://localhost:8080/api/wallets",
      expect.objectContaining({ method: "POST", body: JSON.stringify({ name: "demo" }) })
    );
  });

  it("lists wallets", async () => {
    mockResponse(200, [{ id: "w1", name: "demo", status: "active", created_at: "2026-01-01T00:00:00Z" }]);
    const wallets = await client.listWallets();
    expect(wallets).toHaveLength(1);
    expect(wallets[0].id).toBe("w1");
  });

  it("gets balance", async () => {
    mockResponse(200, [{ chain_id: "eip155:8453", token_address: "0x0", balance: "1000", decimals: 6 }]);
    const balances = await client.getBalance("w1");
    expect(balances[0].balance).toBe("1000");
  });

  it("funds a wallet", async () => {
    mockResponse(200, { txHash: "0xabc" });
    const res = await client.fundWallet("w1", "0xto", "1", "base");
    expect(res.txHash).toBe("0xabc");
  });

  it("fetches swap quote", async () => {
    mockResponse(200, {
      from_token: "0xA",
      to_token: "0xB",
      from_amount: "1000",
      to_amount: "2000",
      chain: "base",
    });
    const quote = await client.swapQuote("w1", {
      fromToken: "0xA",
      toToken: "0xB",
      amount: "1000",
      chain: "base",
    });
    expect(quote.to_amount).toBe("2000");
  });

  it("generates ai text", async () => {
    mockResponse(200, { text: "hello", cost: "0.001" });
    const res = await client.aiGenerate({ walletId: "w1", model: "claude", prompt: "hi" });
    expect(res.text).toBe("hello");
  });

  it("throws GradienceError on non-ok response", async () => {
    mockResponse(400, { error: "bad request" });
    try {
      await client.createWallet("");
    } catch (e: any) {
      expect(e).toBeInstanceOf(GradienceError);
      expect(e.message).toBe("bad request");
      expect(e.statusCode).toBe(400);
    }
  });

  it("exports audit logs", async () => {
    mockResponse(200, [{ id: "a1", action: "sign" }]);
    const logs = await client.exportAudit("w1", "json");
    expect(Array.isArray(logs)).toBe(true);
  });
});

File: sdk/go/gradience/client_test.go
package gradience

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func setupTestServer() (*httptest.Server, *Client) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")

		switch r.URL.Path {
		case "/api/wallets":
			if r.Method == "POST" {
				json.NewEncoder(w).Encode(map[string]string{"id": "w1", "name": "demo", "status": "active", "created_at": "2026-01-01T00:00:00Z"})
			} else {
				json.NewEncoder(w).Encode([]map[string]string{{"id": "w1", "name": "demo", "status": "active", "created_at": "2026-01-01T00:00:00Z"}})
			}
		case "/api/wallets/w1/balance":
			json.NewEncoder(w).Encode([]Balance{{ChainID: "eip155:8453", TokenAddress: "0x0", Balance: "1000", Decimals: 6}})
		case "/api/wallets/w1/fund":
			json.NewEncoder(w).Encode(map[string]string{"txHash": "0xabc"})
		case "/api/swap/quote":
			json.NewEncoder(w).Encode(map[string]string{"from_token": "0xA", "to_token": "0xB", "from_amount": "1000", "to_amount": "2000", "chain": "base"})
		case "/api/ai/generate":
			json.NewEncoder(w).Encode(map[string]string{"text": "hello", "cost": "0.001"})
		default:
			w.WriteHeader(http.StatusNotFound)
			json.NewEncoder(w).Encode(map[string]string{"error": "not found"})
		}
	}))

	client := NewClient(server.URL, "test-token")
	return server, client
}

func TestCreateWallet(t *testing.T) {
	server, client := setupTestServer()
	defer server.Close()

	wallet, err := client.CreateWallet("demo")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if wallet.ID != "w1" {
		t.Errorf("expected id w1, got %s", wallet.ID)
	}
}

func TestListWallets(t *testing.T) {
	server, client := setupTestServer()
	defer server.Close()

	wallets, err := client.ListWallets()
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(wallets) != 1 {
		t.Errorf("expected 1 wallet, got %d", len(wallets))
	}
}

func TestGetBalance(t *testing.T) {
	server, client := setupTestServer()
	defer server.Close()

	balances, err := client.GetBalance("w1")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if balances[0].Balance != "1000" {
		t.Errorf("expected balance 1000, got %s", balances[0].Balance)
	}
}

func TestFundWallet(t *testing.T) {
	server, client := setupTestServer()
	defer server.Close()

	res, err := client.FundWallet("w1", "0xto", "1", "base")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if res["txHash"] != "0xabc" {
		t.Errorf("unexpected txHash: %v", res["txHash"])
	}
}

func TestSwapQuote(t *testing.T) {
	server, client := setupTestServer()
	defer server.Close()

	res, err := client.SwapQuote("w1", map[string]string{
		"from_token": "0xA",
		"to_token":   "0xB",
		"amount":     "1000",
		"chain":      "base",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if res["to_amount"] != "2000" {
		t.Errorf("unexpected to_amount: %v", res["to_amount"])
	}
}

func TestAIGenerate(t *testing.T) {
	server, client := setupTestServer()
	defer server.Close()

	res, err := client.AIGenerate("w1", "claude", "hi")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if res["text"] != "hello" {
		t.Errorf("unexpected text: %v", res["text"])
	}
}
